pkg/config/structure/unmarshal.go

```diff
@@ -32,19 +36,46 @@ func UnmarshalKey(cfg model.Reader, key string, target interface{}) error {
 	}
 	switch outValue.Kind() {
 	case reflect.Map:
-		return copyMap(outValue, source)
+		return copyMap(outValue, source, fs)
 	case reflect.Struct:
-		return copyStruct(outValue, source)
+		return copyStruct(outValue, source, fs)
 	case reflect.Slice:
 		if arr, ok := source.(arrayNode); ok {
-			return copyList(outValue, arr)
+			return copyList(outValue, arr, fs)
+		}
+		if isEmptyString(source) {
+			if fs.convertEmptyStrNil {
+				return nil
+			}
+			return fmt.Errorf("treating empty string as a nil slice not allowed for UnmarshalKey without ConvertEmptyStrNil option")
 		}
-		return fmt.Errorf("can not UnmarshalKey to a slice from a non-list source")
+		return fmt.Errorf("can not UnmarshalKey to a slice from a non-list source: %T", source)
 	default:
 		return fmt.Errorf("can only UnmarshalKey to struct, map, or slice, got %v", outValue.Kind())
 	}
 }
 
+// features allowed for handling edge-cases
+
+type featureSet struct {
+	allowSquash        bool
+	convertEmptyStrNil bool
+}
+
+// UnmarshalKeyOption is an option that affects the enabled features in UnmarshalKey
+type UnmarshalKeyOption func(*featureSet)
+
+// EnableSquash allows UnmarshalKey to take advantage of `mapstructure`s `squash` feature
+// a squashed field hoists its fields up a level in the marshalled representation and directly embeds them
+var EnableSquash UnmarshalKeyOption = func(fs *featureSet) {
+	fs.allowSquash = true
+}
+
+// ConvertEmptyStringToNil allows UnmarshalKey to implicitly convert empty strings into nil slices
+var ConvertEmptyStringToNil UnmarshalKeyOption = func(fs *featureSet) {
+	fs.convertEmptyStrNil = true
+}
+
```

Comment:
Nit: Should we move it above the `UnmarshalKey` function definition? 

I always find easier to check for type definitions and `var` definitions in the top level of the file or above the function they are used.